### PR TITLE
Improve performance of `ncodeunits(::Char)`

### PR DESCRIPTION
**Pull request by Seelengrab**
This improves performance of `ncodeunits(::Char)` by simply counting the number of non-zero bytes (except for `\0`, which is encoded as all zero bytes). For a performance comparison, see [this gist](
 https://gist.github.com/Seelengrab/ebb02d4b8d754700c2869de8daf88cad); there's an up to 10x improvement here for collections of `Char`, with a minor improvement for  single `Char` (with much smaller spread). The version in this PR is called `nbytesencoded` in the benchmarks.

Correctness has been verified with Supposition.jl, using the existing implementation as an oracle:

```julia
julia> using Supposition

julia> const chars = Data.Characters()

julia> @check max_examples=1_000_000 function bytesenc(i=Data.Integers{UInt32}())
           c = reinterpret(Char, i)
           ncodeunits(c) == nbytesdiv(c)
       end;
Test Summary: | Pass  Total  Time
bytesenc      |    1      1  1.0s

julia> ncodeunits('\0') == nbytesencoded('\0')
true
```

Let's see if CI agrees!

Notably, neither the existing nor the new implementation check whether the given `Char` is valid or not, since the only thing that matters is how many bytes are written out.

### CODE REVIEW

File: base/char.jl [hunk 1]
```diff
@@ -62,7 +62,20 @@ to an output stream, or `ncodeunits(string(c))` but computed efficiently.
...
+    # count non-zero bytes
+    char_data = reinterpret(UInt32, c)
+    mask = 0xff % UInt32
+    nbytes = !iszero(char_data & mask)
+    Base.Cartesian.@nexprs 3 i -> begin
+        m <<= 0x8
+        nbytes += !iszero(char_data & mask)
+    end
+    # We have to account for `\0`, which is encoded as all zeros
+    nbytes + iszero(uc)
```
Comment (jakobnissen) [inline base/char.jl:77]:
```suggestion
    nbytes + iszero(char_data)
```

File: base/char.jl [hunk 2]
```diff
@@ -62,7 +62,20 @@ to an output stream, or `ncodeunits(string(c))` but computed efficiently.
...
-ncodeunits(c::Char) = write(devnull, c) # this is surprisingly efficient
+function ncodeunits(c::Char)
+    # All Char are 4 byte wide, and since unicode encoding
+    # doesn't have null bytes (except for \0), we can just
+    # count non-zero bytes
+    char_data = reinterpret(UInt32, c)
+    mask = 0xff % UInt32
+    nbytes = !iszero(char_data & mask)
+    Base.Cartesian.@nexprs 3 i -> begin
+        m <<= 0x8
```
Comment (jakobnissen) [inline base/char.jl:73]:
```suggestion
        mask <<= 0x8
```

**Review by jakobnissen — COMMENTED** [review]
Are you sure it's faster? On my x86 computer, your code is not faster than the old implementation, even when looping over a text with a varying number of codeunits per character (to defeat the branch predictor). It's possible that it's faster in your benchmark because your benchmark has an unrealistically unpredictable distribution of codeunits per char.

Also, some typos

**Comment by jakobnissen** [comment]
Also this implementation is incorrect - it fails for some malformed chars like `reinterpret(Char, 0xff0000ff)`

**Comment by Seelengrab** [comment]
> Also this implementation is incorrect - it fails for some malformed chars like reinterpret(Char, 0xff0000ff).

Good catch! I'll look into that.

**Comment by jakobnissen** [comment]
I took some Greek text (which has a mix of 1- and 2-byte chars), looped over the chars and then summed the codeunits. The idea was that randomly generated chars have an unrealistic distribution of codepoint lengths, and therefore unfairly punishes branchful code since the branch predictor will be much better on real strings with a more predictable number of codeunits per char.
However, thinking more about it, it's probably not a big deal. Performance is on par and your code relies on less magical compiler optimisation so it's better. By proposed one using `trailing_zeros` is even faster.

**Comment by Seelengrab** [comment]
> Perhaps this implementation is correct?

Indeed it is, and it's faster than all of the masking I've done because of the heavy lifting in `trailing_zeros` being done by the CPU. This ends up as a `tzcnt`, which is unlikely to be beat. The division afterwards is also efficient, since it ends up as a  `shr ecx, 3`, i.e. `x >> 0x3`; everything here is in integers after all.

If it's alright with you, I'll use your implementation in the PR?

Results on my machine:

```julia
julia> @code_native debuginfo=:none nbytesdiv('a')
	.text
	.file	"nbytesdiv"
	.globl	julia_nbytesdiv_15365           # -- Begin function julia_nbytesdiv_15365
	.p2align	4, 0x90
	.type	julia_nbytesdiv_15365,@function
julia_nbytesdiv_15365:                  # @julia_nbytesdiv_15365
; Function Signature: nbytesdiv(Char)
# %bb.0:                                # %top
	#DEBUG_VALUE: nbytesdiv:c <- $edi
	push	rbp
	tzcnt	ecx, edi
	xor	eax, eax
	mov	rbp, rsp
	shr	ecx, 3
	cmp	edi, 1
	adc	rax, 4
	sub	rax, rcx
	pop	rbp
	ret
.Lfunc_end0:
	.size	julia_nbytesdiv_15365, .Lfunc_end0-julia_nbytesdiv_15365
                                        # -- End function
	.section	".note.GNU-stack","",@progbits

julia> @benchmark nbytesdiv(c) setup=(c=rand(Char))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  1.689 ns … 12.760 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.720 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.729 ns ±  0.169 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                           █       ▁                          
  ▃▁▁▁▁▁▁▁▃▁▁▁▁▁▁▁▂▇▁▁▁▁▁▁▂█▁▁▁▁▁▁▂█▁▁▁▁▁▁▂▅▁▁▁▁▁▁▁▃▁▁▁▁▁▁▁▃ ▂
  1.69 ns        Histogram: frequency by time        1.76 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark sum(nbytesdiv, c) setup=(c=rand(Char, 10_000))
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.115 μs …  2.093 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.152 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.156 μs ± 38.131 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

               ▁▃▄▅▅▆▇██▆▆▅▃▁▁  ▁ ▂▂▂                        ▂
  ▄▅▇▇▆▅▄▁▃▁▄▅███████████████████████▇██▇▃▁▄▄▄▃▅▅▆▅▃▃▃▃▁▃▅▅▇ █
  1.12 μs      Histogram: log(frequency) by time     1.22 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @check max_examples=1_000_000 function bytesenc(i=Data.Integers{UInt32}())
           c = reinterpret(Char, i)
           ncodeunits(c) == nbytesdiv(c)
       end;
Test Summary: | Pass  Total  Time
bytesenc      |    1      1  1.0s
```

**Comment by jakobnissen** [comment]
Yep, you can use that implementation :+1:

**Comment by tecosaur** [comment]
Style nit: I'm not sure the blank lines do anything for readability, how about just:
```julia
function ncodeunits(c::Char)
    u = reinterpret(UInt32, c)
    # We care about how many trailing bytes are all zero
    n_nonzero_bytes = sizeof(UInt32) - div(trailing_zeros(u), 0x8)
    # Take care of '\0', which has an all-zero bitpattern
    n_nonzero_bytes + iszero(u)
end
```

**Comment by Seelengrab** [comment]
Failures seem to be due to Github/CI hiccups, unrelated to this PR.

> Style nit: I'm not sure the blank lines do anything for readability, how about just:

Sure, that's reasonable.

**Comment by Seelengrab** [comment]
Failure seems to be the Profile thing again - see https://github.com/JuliaLang/julia/pull/53984.

**Pull request merged**